src/Features/LanguageServer/Protocol/Handler/MapCode/MapCodeHandler.cs

```
@@ -0,0 +1,137 @@
+        {
+            return new WorkspaceEdit
+            {
+                DocumentChanges = uriToEditsMap.Select(kvp => new TextDocumentEdit
+                {
+                    TextDocument = new OptionalVersionedTextDocumentIdentifier { Uri = kvp.Key },
+                    Edits = kvp.Value,
+                }).ToArray()
+            };
+        }
+        else
+        {
+            return new WorkspaceEdit
+            {
+                Changes = uriToEditsMap.ToDictionary(kvp => ProtocolConversions.GetDocumentFilePathFromUri(kvp.Key), kvp => kvp.Value)
+            };
+        }
+
+        async Task<(Uri, TextEdit[])?> MapCodeAsync(LSP.MapCodeMapping codeMapping)
+        {
+            var textDocument = codeMapping.TextDocument;
+            if (textDocument == null)
+                return null;
+
+            if (context.Solution.GetDocument(textDocument) is not Document document)
+                return null;
+
+            var codeMapper = document.GetLanguageService<IMapCodeService>();
+            if (codeMapper == null)
+                return null;
+
+            var focusLocations = await ConvertFocusLocationsToDocumentAndSpansAsync(
+                document,
+                textDocument,
+                codeMapping.FocusLocations,
+                cancellationToken).ConfigureAwait(false);
+
+            var newDocument = await codeMapper.MapCodeAsync(
+                document,
+                codeMapping.Contents.ToImmutableArrayOrEmpty(),
+                focusLocations,
+                formatMappedCode: true,
+                cancellationToken).ConfigureAwait(false);
+
+            if (newDocument is null)
+                return null;
+
+            var textChanges = (await newDocument.GetTextChangesAsync(document, cancellationToken).ConfigureAwait(false)).ToImmutableArray();
+            if (textChanges.IsEmpty)
+                return null;
```

If the mapper did not map anything, does it return the unchanged document, or null?  If it is the latter then this should throw as an empty set of changes would be unexpected

---

the way the mapCode request is defined today, it can include multiple map requests, so if we failed to map one of them, we'd still try provide results for others. I will log a warning here